### PR TITLE
Initial round of improvements to volumeToMesh

### PR DESCRIPTION
**Pull request by Idclip**
A very quick round of improvements/modernisations to volumeToMesh. Tested on many of the typical vdb cached level sets (crawler/iss/etc). 

Results in 10-15% improved performance.

Tested on a M1 Max (64GB/10CPU/system malloc/AppleClang 13.1.6)
Tested on Ubuntu x86/64 (64GB/12CPU E5-1650 v4 @ 3.60GHz/jemalloc 5/GCC 7.5)

There are many more additional (but more complicated) improvements that can be made in future PRs.

Signed-off-by: Nick Avramoussis <4256455+Idclip@users.noreply.github.com>

### CODE REVIEW

File: openvdb/openvdb/tools/VolumeToMesh.h [hunk 1]
```diff
@@ -608,57 +608,94 @@ inline bool isBoolValue() { return false; }
...
-
 template<typename T>
 inline bool isInsideValue(T value, T isovalue) { return value < isovalue; }
 
 template<>
 inline bool isInsideValue<bool>(bool value, bool /*isovalue*/) { return value; }
 
 
-template<typename AccessorT>
+/// @brief  Minor wrapper around the Leaf API to avoid atmoic access with
```
Comment (danrbailey) [inline openvdb/openvdb/tools/VolumeToMesh.h:618]: Good idea!

typo: atomic

File: openvdb/openvdb/tools/VolumeToMesh.h [hunk 2]
```diff
@@ -608,57 +608,94 @@ inline bool isBoolValue() { return false; }
...
+    using T = bool;
+    LeafBufferAccessor(const LeafT& leaf) : mLeaf(leaf) {}
+    inline T get(const Index idx) const { return mLeaf.getValue(idx); }
+    const LeafT& mLeaf;
+};
+
+
+/// @brief  Whether a coordinate does not lie at the positive edge of a leaf node.
+template <typename LeafT>
+inline bool isInternalLeafCoord(const Coord& ijk)
```
Comment (danrbailey) [inline openvdb/openvdb/tools/VolumeToMesh.h:642]: Remove redundant inline keyword on templated functions.

File: openvdb/openvdb/tools/VolumeToMesh.h [hunk 3]
```diff
@@ -948,29 +971,28 @@ mergeVoxels(LeafType& leaf, const Coord& start, int dim, int regionId)
 // thinking that it is a constructor.
 template <class LeafType>
 inline bool
-isMergable(LeafType& leaf, const Coord& start, int dim,
+isMergable(const LeafType& leaf,
```
Comment (danrbailey) [inline openvdb/openvdb/tools/VolumeToMesh.h:974]: Might as well fix the typo here - isMergeable

File: openvdb/openvdb/tools/VolumeToMesh.h [hunk 4]
```diff
@@ -3137,44 +3105,51 @@ evalInternalVoxelEdges(VoxelEdgeAcc& edgeAcc, const LeafNode& leafnode,
 /// Utility method to check for sign changes along the x + 1, y + 1 or z + 1 directions.
 /// The direction is determined by the @a edgeAcc parameter. All voxels that reside in the
 /// specified leafnode face: back, top or right are evaluated.
-template<typename LeafNode, typename TreeAcc, typename VoxelEdgeAcc>
+template<typename LeafNodeT, typename TreeAcc, typename VoxelEdgeAcc>
 void
-evalExtrenalVoxelEdges(VoxelEdgeAcc& edgeAcc, TreeAcc& acc, const LeafNode& lhsNode,
-    const LeafNodeVoxelOffsets& voxels, const typename LeafNode::ValueType iso)
+evalExtrenalVoxelEdges(VoxelEdgeAcc& edgeAcc,
```
Comment (danrbailey) [inline openvdb/openvdb/tools/VolumeToMesh.h:3110]: Another typo here worth fixing up - "External"

File: openvdb/openvdb/tools/VolumeToMesh.h [hunk 5]
```diff
@@ -3185,29 +3160,34 @@ evalExtrenalVoxelEdges(VoxelEdgeAcc& edgeAcc, TreeAcc& acc, const LeafNode& lhsN
 /// Utility method to check for sign changes along the x - 1, y - 1 or z - 1 directions.
 /// The direction is determined by the @a edgeAcc parameter. All voxels that reside in the
 /// specified leafnode face: front, bottom or left are evaluated.
-template<typename LeafNode, typename TreeAcc, typename VoxelEdgeAcc>
+template<typename LeafNodeT, typename TreeAcc, typename VoxelEdgeAcc>
 void
-evalExtrenalVoxelEdgesInv(VoxelEdgeAcc& edgeAcc, TreeAcc& acc, const LeafNode& leafnode,
-    const LeafNodeVoxelOffsets& voxels, const typename LeafNode::ValueType iso)
+evalExtrenalVoxelEdgesInv(VoxelEdgeAcc& edgeAcc,
```
Comment (danrbailey) [inline openvdb/openvdb/tools/VolumeToMesh.h:3165]: "External"

File: openvdb/openvdb/tools/VolumeToMesh.h [hunk 6]
```diff
@@ -4648,6 +4607,65 @@ relaxDisorientedTriangles(
 }
 
 
+template<typename GridType>
+inline void
```
Comment (danrbailey) [inline openvdb/openvdb/tools/VolumeToMesh.h:4611]: Redundant inline

**Review by danrbailey — APPROVED** [review]
This is awesome, hard to overstate how impactful changes like these are. The sheer amount of compute time spent in these routines is huge as I'm sure you're aware.

Out of interest, do you happen to know where the majority of the performance boost comes from. I'm guessing it's either eliminating the atomic access on calls to `getValue()` or switching from `std::vector` to `std::array`?

If it's the atomic access, makes me think we could perhaps check if delayed loading is disabled in the `LeafBuffer` to also skip the atomic access. :/

**Comment by Idclip** [comment]
> Out of interest, do you happen to know where the majority of the performance boost comes from. I'm guessing it's either eliminating the atomic access on calls to getValue() or switching from std::vector to std::array?
> If it's the atomic access, makes me think we could perhaps check if delayed loading is disabled in the LeafBuffer to also skip the atomic access. :/

I think it's primarily swapping from those small heap allocated vectors to stack arrays. That being said the atomic stuff is a real problem. The fact we have to pay this price for intermediate (non-serialized) grid data is very sad:

https://godbolt.org/z/4M1Ee71sK
https://quick-bench.com/q/C9sLRkfzKj72nX2O-3TDl4pEDpk

I don't have a good solution - it's almost as if we need two grid types e.g:

```
template<typename T, Index N1=5, Index N2=4, Index N3=3>
struct Tree4 {
    using Type = Tree<RootNode<InternalNode<InternalNode<LeafNode<T, N3, /*use_atmoic*/true>, N2>, N1>>>;
};

namespace internal {
template<typename T, Index N1=5, Index N2=4, Index N3=3>
struct Tree4NonSerializable {
    using Type = Tree<RootNode<InternalNode<InternalNode<LeafNode<T, N3, /*use_atmoic*/false>, N2>, N1>>>;
};
}

using FloatTree = Tree4<float>;
using FloatTreeInMem = internal::Tree4NonSerializable<float>; // This is an internal type that's only used in 
```

It's not a massive problem internally because we design our algorithms (generally) to pull out the raw buffer data instead of going through the Leaf API. But anything that uses the value iterators pays this price. Another option could be to change the ValueIterators on the LeafNodes to something like:

https://github.com/AcademySoftwareFoundation/openvdb/blob/1364816e746284b3f57b25853ea8a3f1d8ed8aad/openvdb/openvdb/tree/LeafNode.h#L213-L246


```
template<typename MaskIterT, typename NodeT, typename ValueT, typename TagT>
struct ValueIter:
    public SparseIteratorBase<
        MaskIterT, ValueIter<MaskIterT, NodeT, ValueT, TagT>, NodeT, ValueT>
{
    ValueIter(const MaskIterT& iter, NodeT* parent) 
        : BaseT(iter, parent)
        , mData(parent->buffer().data()) // potentially load and cache the buffer ptr
        {}

    ValueT& getItem(Index pos) const { return this->mData[pos]; }
    ValueT& getValue() const { return this->mData[this->pos()]; }

    void setItem(Index pos, const ValueT& value) const {
        this->mData[this->pos()] = value;
        this->parent().setValueOnly(pos, value);
    }
    void setValue(const ValueT& value) const {
        this->mData[this->pos()] = value;
    }

private:
    float* mData;
}
```

But I haven't considered all the potential consequences of a change like this and of course you still pay for the extra size on the leaf nodes and going through the Leaf API.

File: openvdb/openvdb/tools/VolumeToMesh.h (hunk 1 above)
Comment (Idclip) [reply openvdb/openvdb/tools/VolumeToMesh.h]: Done

File: openvdb/openvdb/tools/VolumeToMesh.h (hunk 3 above)
Comment (Idclip) [reply openvdb/openvdb/tools/VolumeToMesh.h]: Done

File: openvdb/openvdb/tools/VolumeToMesh.h (hunk 4 above)
Comment (Idclip) [reply openvdb/openvdb/tools/VolumeToMesh.h]: Done

**Comment by nachovizzo** [comment]
NOTE: This MR breaks (at least my) build for PyOpenVDB : with `openvdb/tools/VolumeToMesh.h:4621:65: error: static assertion failed: volume to mesh conversion is supported only for scalar grids`

```sh
Building CXX object openvdb/openvdb/python/CMakeFiles/pyopenvdb.dir/pyVec3Grid.cc.o
In file included from /home/ivizzo/dev/libs/OpenVDB/openvdb/openvdb/python/pyGrid.h:19,
                 from /home/ivizzo/dev/libs/OpenVDB/openvdb/openvdb/python/pyVec3Grid.cc:7:
/home/ivizzo/dev/libs/OpenVDB/openvdb/openvdb/../openvdb/tools/VolumeToMesh.h: In instantiation of ‘void openvdb::v9_1::tools::volume_to_mesh_internal::doVolumeToMesh(const GridType&, std::vector<openvdb::v9_1::math::Vec3<float> >&, std::vector<openvdb::v9_1::math::Vec3<unsigned int> >&, std::vector<openvdb::v9_1::math::Vec4<unsigned int> >&, double, double, bool) [with GridType = openvdb::v9_1::Grid<openvdb::v9_1::tree::Tree<openvdb::v9_1::tree::RootNode<openvdb::v9_1::tree::InternalNode<openvdb::v9_1::tree::InternalNode<openvdb::v9_1::tree::LeafNode<openvdb::v9_1::math::Vec3<float>, 3>, 4>, 5> > > >]’:
/home/ivizzo/dev/libs/OpenVDB/openvdb/openvdb/../openvdb/tools/VolumeToMesh.h:5192:44:   required from ‘void openvdb::v9_1::tools::volumeToMesh(const GridType&, std::vector<openvdb::v9_1::math::Vec3<float> >&, std::vector<openvdb::v9_1::math::Vec3<unsigned int> >&, std::vector<openvdb::v9_1::math::Vec4<unsigned int> >&, double, double, bool) [with GridType = openvdb::v9_1::Grid<openvdb::v9_1::tree::Tree<openvdb::v9_1::tree::RootNode<openvdb::v9_1::tree::InternalNode<openvdb::v9_1::tree::InternalNode<openvdb::v9_1::tree::LeafNode<openvdb::v9_1::math::Vec3<float>, 3>, 4>, 5> > > >]’
/home/ivizzo/dev/libs/OpenVDB/openvdb/openvdb/python/pyGrid.h:1469:24:   required from ‘boost::python::api::object pyGrid::volumeToMesh(const GridType&, boost::python::api::object, boost::python::api::object) [with GridType = openvdb::v9_1::Grid<openvdb::v9_1::tree::Tree<openvdb::v9_1::tree::RootNode<openvdb::v9_1::tree::InternalNode<openvdb::v9_1::tree::InternalNode<openvdb::v9_1::tree::LeafNode<openvdb::v9_1::math::Vec3<float>, 3>, 4>, 5> > > >]’
/home/ivizzo/dev/libs/OpenVDB/openvdb/openvdb/python/pyGrid.h:2393:17:   required from ‘void pyGrid::exportGrid() [with GridType = openvdb::v9_1::Grid<openvdb::v9_1::tree::Tree<openvdb::v9_1::tree::RootNode<openvdb::v9_1::tree::InternalNode<openvdb::v9_1::tree::InternalNode<openvdb::v9_1::tree::LeafNode<openvdb::v9_1::math::Vec3<float>, 3>, 4>, 5> > > >]’
/home/ivizzo/dev/libs/OpenVDB/openvdb/openvdb/python/pyVec3Grid.cc:16:34:   required from here
/home/ivizzo/dev/libs/OpenVDB/openvdb/openvdb/../openvdb/tools/VolumeToMesh.h:4621:65: error: static assertion failed: volume to mesh conversion is supported only for scalar grids
 4621 |     static_assert(std::is_scalar<typename GridType::ValueType>::value,
      |                                                                 ^~~~~
/home/ivizzo/dev/libs/OpenVDB/openvdb/openvdb/../openvdb/tools/VolumeToMesh.h:4621:65: note: ‘std::integral_constant<bool, false>::value’ evaluates to false
make[2]: *** [openvdb/openvdb/python/CMakeFiles/pyopenvdb.dir/build.make:160: openvdb/openvdb/python/CMakeFiles/pyopenvdb.dir/pyVec3Grid.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:230: openvdb/openvdb/python/CMakeFiles/pyopenvdb.dir/all] Error 2

```

I guess it's because pyVec3Grid is "asking" for that

**Comment by Idclip** [comment]
Argh this is only hit with Numpy enabled which I don't think we test in our CI (@danrbailey fyi, more CI) - thanks @nachovizzo, I'll fix this asap

**Pull request merged**